### PR TITLE
IPN verification

### PR DESCRIPTION
**Pull request by KaiNissen**

### CODE REVIEW

File: app/config/config.dist.json [hunk 1]
```diff
@@ -90,7 +90,10 @@
 		"notify-url": "",
 		"return-url": "",
 		"cancel-url": "",
-		"item-name": ""
+		"item-name": "",
+		"confirmable-statuses": [
+			"Completed"
+		]
```
Comment (JeroenDeDauw) [inline app/config/config.dist.json]: Is this used? And why is it in the config?

File: src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @license GNU GPL v2+
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ */
+class HandlePayPalPaymentNotificationUseCase {
+
+	private $handler;
+
+	public function __construct( PayPalPaymentNotificationHandler $handler ) {
```
Comment (JeroenDeDauw) [inline src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php]: Will this UC have more than just this handler? I'm wondering what the difference is

Comment (KaiNissen) [inline src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php]: Right, actually there is no handler class needed.

File: app/config/config.dist.json (hunk 1 above)
Comment (KaiNissen) [inline app/config/config.dist.json]: It's used now, but I removed it from the config.

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ */
+class PayPalPaymentNotificationVerifier {
+
+	/** @var Client */
+	private $httpClient;
+	private $config;
+	private $allowedStatuses = [ 'Completed' ];
+
+	public function __construct( $httpClient, array $config ) {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: No type hint in the constructor?

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		}
+
+		if ( trim( $result->getBody()->getContents() ) !== 'VERIFIED' ) {
+			throw new PayPalPaymentNotificationVerifierException( 'Payment provider did not confirm the sent data' );
+		}
+
+		return true;
+	}
+
+	private function matchReceiverAddress( $request ) {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: private function matchesReceiverAddress( $request ): bool {

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	/**
+	 * Verifies the request's integrity and reassures with PayPal
+	 * servers that the request wasn't tampered with during transfer
+	 *
+	 * @param array $request
+	 *
+	 * @return bool
+	 * @throws PayPalPaymentNotificationVerifierException
+	 */
+	public function verify( array $request ): bool {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: return value not needed and not used

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		}
+
+		if ( trim( $result->getBody()->getContents() ) !== 'VERIFIED' ) {
+			throw new PayPalPaymentNotificationVerifierException( 'Payment provider did not confirm the sent data' );
+		}
+
+		return true;
+	}
+
+	private function matchReceiverAddress( $request ): bool {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: matches

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	private $httpClient;
+	private $config;
+	private $allowedStatuses = [ 'Completed' ];
+
+	/**
+	 * PayPalPaymentNotificationVerifier constructor.
+	 * @param Client $httpClient
+	 * @param array $config
+	 */
+	public function __construct( $httpClient, array $config ) {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: type hint in code missing

File: tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\Tests\System\Routes;
+
+use GuzzleHttp\Psr7\Response;
+use GuzzleHttp\Psr7\Stream;
+use Symfony\Component\HttpKernel\Client;
```
Comment (JeroenDeDauw) [inline tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php]: Should be guzzle client?

File: tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+					'payment_status' => 'Unknown'
+				]
+			);
+
+			$this->assertSame( 'TODO', $client->getResponse()->getContent() );
+			$this->assertSame( 200, $client->getResponse()->getStatusCode() );
+		} );
+	}
+
+	private function getClientMock() {
```
Comment (JeroenDeDauw) [inline tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php]: return type hint missing

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+			->getMock();
+
+		$body->expects( $this->once() )
+			->method( 'getContents' )
+			->willReturn( 'INVALID' );
+
+		return $this->newClient( $body );
+	}
+
+	private function newClient( $body ) {
```
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: return type hint

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+			->getMock();
+
+		$body->expects( $this->once() )
+			->method( 'getContents' )
+			->willReturn( 'VERIFIED' );
+
+		return $this->newClient( $body );
+	}
+
+	private function newFailingClient() {
```
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: return type hint

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+		return new PayPalPaymentNotificationVerifier(
+			$client,
+			[
+				'base-url' => 'https://dummy-url.com',
+				'account-address' => 'foerderpp@wikimedia.de'
+			]
+		);
+	}
+
+	private function newSucceedingClient() {
```
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: return type hint

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php (hunk 10 above)
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: param type hint

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+	public function testReassuringReceivedDataFails_verifierReturnsFalse() {
+		$this->expectException( PayPalPaymentNotificationVerifierException::class );
+		$verifier = $this->newVerifier( $this->newFailingClient() );
+		$verifier->verify( [
+			'receiver_email' => 'foerderpp@wikimedia.de',
+			'payment_status' => 'Completed'
+		] );
+	}
+
+	private function newVerifier( $client ) {
```
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: Client $httpClient

File: tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+	public function testReceiverAddressNotGiven_verifierReturnsFalse() {
+		$this->expectException( PayPalPaymentNotificationVerifierException::class );
+
+		$this->newVerifier( null )->verify( [] );
+	}
+
+	public function testPaymentStatusNotGiven_verifierReturnsFalse() {
+		$this->expectException( PayPalPaymentNotificationVerifierException::class );
+		$this->newVerifier( null )->verify( [
+			'receiver_email' => 'foerderpp@wikimedia.de'
```
Comment (JeroenDeDauw) [inline tests/Unit/Infrastructure/PayPalPaymentNotificationVerifierTest.php]: extract email to constant

File: tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+			->getMock();
+
+		$client->expects( $this->any() )
+			->method( 'post' )
+			->with(
+				'https://that.paymentprovider.com/',
+				[
+					'cmd' => '_notify_validate',
+					'receiver_email' => 'foerderpp@wikimedia.de',
+					'payment_status' => 'Completed'
```
Comment (JeroenDeDauw) [inline tests/System/Routes/HandlePayPalPaymentNotificationRouteTest.php]: extract email and url to constants

File: app/routes.php [hunk 16]
```diff
@@ -375,4 +378,63 @@ function( Application $app, Request $request ) use ( $ffFactory ) {
...
+				} catch ( PayPalPaymentNotificationVerifierException $e ) {
+					// TODO: log error
+				}
+
+				return new Response( 'TODO' ); # PayPal expects an empty response
+			}
+
+			private function getTokenFromCustomVars( string $customVars ): string {
+				$vars = json_decode( $customVars, true );
+				if ( !empty( $customVars['token'] ) ) {
```
Comment (JeroenDeDauw) [inline app/routes.php]: should be `$vars['token']`?

**Pull request merged**